bmr: Error allocating static memory on aarch64

While trying to run the tests on FreeBSD 13.0-ALPHA1, I see:

```
[nico@henricus ~/src/bm]$ ./run\-tests.sh 
+ ./build.sh
+ CC=/usr/bin/cc
+ CFLAGS='-Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb -std=c11'
+ LIBS=''
+ mkdir -p build/bin
+ mkdir -p build/examples
+ /usr/bin/cc -Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb '-std=c11' -o build/bin/basm ./src/basm.c
+ /usr/bin/cc -Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb '-std=c11' -o build/bin/bme ./src/bme.c
+ /usr/bin/cc -Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb '-std=c11' -o build/bin/bmr ./src/bmr.c
+ /usr/bin/cc -Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb '-std=c11' -o build/bin/debasm ./src/debasm.c
+ /usr/bin/cc -Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb '-std=c11' -o build/bin/bdb ./src/bdb.c
./src/bdb.c:347:10: warning: overflow converting case value to switch condition type (-1 to 255) [-Wswitch]
    case EOF:
         ^
/usr/include/stdio.h:210:13: note: expanded from macro 'EOF'
#define EOF     (-1)
                ^
1 warning generated.
+ /usr/bin/cc -Wall -Wextra -Wswitch-enum -Wmissing-prototypes -Wconversion -pedantic -fno-strict-aliasing -ggdb '-std=c11' -o build/bin/basm2nasm ./src/basm2nasm.c
+ find examples/ -name '*.basm'
+ sed 's/\.basm//'
+ ./build/bin/basm -g examples/cast.basm build/examples/cast.bm
+ ./build/bin/basm -g examples/chars.basm build/examples/chars.bm
+ ./build/bin/basm -g examples/e.basm build/examples/e.bm
+ ./build/bin/basm -g examples/hello.basm build/examples/hello.bm
+ ./build/bin/basm -g examples/123i.basm build/examples/123i.bm
+ ./build/bin/basm -g examples/fib.basm build/examples/fib.bm
+ ./build/bin/basm -g examples/bits.basm build/examples/bits.bm
+ ./build/bin/basm -g examples/123f.basm build/examples/123f.bm
+ ./build/bin/basm -g examples/pi.basm build/examples/pi.bm
+ ./build/bin/basm -g examples/rot13.basm build/examples/rot13.bm
+ ./build/bin/basm -g examples/gray.basm build/examples/gray.bm
+ find examples/ -name '*.basm'
+ sed 's/\.basm//'
+ ./build/bin/bmr -p build/examples/cast.bm -eo test/examples/cast.expected.out
Data segment size exceeds process limit
Abort trap
[nico@henricus ~/src/bm]$ uname -a
FreeBSD henricus.herrhotzenplotz.geek 13.0-ALPHA1 FreeBSD 13.0-ALPHA1 #0 b9fdd1446: Wed Jan 20 17:39:57 CET 2021     nico@triton.herrhotzenplotz.geek:/usr/obj/usr/src/arm64.aarch64/sys/GENERIC-NODEBUG  arm64
[nico@henricus ~/src/bm]$ 
```


Have fun with this one ;-)

--- a/nobuild.c
+++ b/nobuild.c
@@ -2,7 +2,7 @@
 #include "./nobuild.h"
 
 #ifdef _WIN32
-#define CFLAGS "/std:c11", "/O2", "/FC", "/W4", "/WX", "/wd4996", "/nologo", "/Fe.\\build\\bin\\", "/Fo.\\build\\bin\\"
+#define CFLAGS "/std:c11", "/O2", "/FC", "/W4", "/WX", "/wd4996", "/wd4200", "/nologo", "/Fe.\\build\\bin\\", "/Fo.\\build\\bin\\"
 #else
 #define CFLAGS "-Wall", "-Wextra", "-Wswitch-enum", "-Wmissing-prototypes", "-Wconversion", "-pedantic", "-fno-strict-aliasing", "-ggdb", "-std=c11"
 #endif

--- a/src/basm.c
+++ b/src/basm.c
@@ -40,7 +40,9 @@ int main(int argc, char **argv)
     const char *output_file_path = shift(&argc, &argv);
 
     // NOTE: The structure might be quite big due its arena. Better allocate it in the static memory.
-    static Basm basm = {0};
+    static Basm basm = {
+        .arena = {.capacity = ARENA_RECOMMENDED_CAPACITY}
+    };
     basm_translate_source(&basm, sv_from_cstr(input_file_path));
 
     if (!basm.has_entry) {
@@ -78,6 +80,8 @@ int main(int argc, char **argv)
         }
         fclose(symbol_file);
     }
+
+    arena_free(&basm.arena);
 
     return 0;
 }

--- a/src/basm2nasm.c
+++ b/src/basm2nasm.c
@@ -39,7 +39,11 @@ int main(int argc, char *argv[])
 
 
     // NOTE: The structure might be quite big due its arena. Better allocate it in the static memory.
-    static Basm basm = {0};
+    static Basm basm = {
+        .arena = {
+            .capacity = ARENA_RECOMMENDED_CAPACITY
+        }
+    };
     basm_translate_source(&basm, sv_from_cstr(input_file_path));
 
     FILE *output = fopen(output_file_path, "wb");
@@ -447,6 +451,8 @@ int main(int argc, char *argv[])
     fprintf(output, "stack: resq BM_STACK_CAPACITY\n");
 
     fclose(output);
+
+    arena_free(&basm.arena);
 
     return 0;
 }

--- a/src/bdb.c
+++ b/src/bdb.c
@@ -513,7 +513,11 @@ int main(int argc, char **argv)
     }
 
     // NOTE: The structure might be quite big due its arena. Better allocate it in the static memory.
-    static Bdb_State state = {0};
+    static Bdb_State state = {
+        .arena = {
+            .capacity = ARENA_RECOMMENDED_CAPACITY
+        }
+    };
     state.bm.halt = 1;
 
     printf("BDB - The birtual machine debugger.\n"
@@ -575,6 +579,8 @@ int main(int argc, char **argv)
             return EXIT_SUCCESS;
         }
     }
+
+    arena_free(&state.arena);
 
     return EXIT_SUCCESS;
 }

--- a/src/bm.h
+++ b/src/bm.h
@@ -216,6 +216,8 @@ struct Region {
 
 Region *region_create(size_t capacity);
 
+#define ARENA_RECOMMENDED_CAPACITY (640 * 1000)
+
 typedef struct {
     size_t capacity;
     Region *first;

--- a/src/bm.h
+++ b/src/bm.h
@@ -32,9 +32,6 @@
 #define BASM_PP_SYMBOL '%'
 #define BASM_MAX_INCLUDE_LEVEL 69
 
-//#define ARENA_CAPACITY (1000 * 1000 * 1000)
-#define ARENA_CAPACITY (640 * 1000)
-
 typedef struct {
     size_t count;
     const char *data;
@@ -208,12 +205,28 @@ PACK(struct Bm_File_Meta {
 typedef struct Bm_File_Meta Bm_File_Meta;
 
 // NOTE: https://en.wikipedia.org/wiki/Region-based_memory_management
-typedef struct {
-    char buffer[ARENA_CAPACITY];
+typedef struct Region Region;
+
+struct Region {
+    Region *next;
+    size_t capacity;
     size_t size;
+    char buffer[];
+};
+
+Region *region_create(size_t capacity);
+
+typedef struct {
+    size_t capacity;
+    Region *first;
+    Region *last;
 } Arena;
 
 void *arena_alloc(Arena *arena, size_t size);
+void arena_clean(Arena *arena);
+void arena_free(Arena *arena);
+void arena_summary(Arena *arena);
+
 int arena_slurp_file(Arena *arena, String_View file_path, String_View *content);
 const char *arena_sv_to_cstr(Arena *arena, String_View sv);
 String_View arena_sv_concat2(Arena *arena, const char *a, const char *b);
@@ -1183,13 +1196,82 @@ String_View arena_sv_concat2(Arena *arena, const char *a, const char *b)
     };
 }
 
+Region *region_create(size_t capacity)
+{
+    const size_t region_size = sizeof(Region) + capacity;
+    Region *region = malloc(region_size);
+    memset(region, 0, region_size);
+    region->capacity = capacity;
+    return region;
+}
+
 void *arena_alloc(Arena *arena, size_t size)
 {
-    assert(arena->size + size <= ARENA_CAPACITY);
+    if (arena->last == NULL) {
+        assert(arena->first == NULL);
 
-    void *result = arena->buffer + arena->size;
-    arena->size += size;
+        Region *region = region_create(
+            size > arena->capacity ? size : arena->capacity);
+
+        arena->last = region;
+        arena->first = region;
+    }
+
+    while (arena->last->size + size > arena->last->capacity &&
+           arena->last->next) {
+        arena->last = arena->last->next;
+    }
+
+    if (arena->last->size + size > arena->last->capacity) {
+        Region *region = region_create(
+            size > arena->capacity ? size : arena->capacity);
+
+        arena->last->next = region;
+        arena->last = region;
+    }
+
+    void *result = arena->last->buffer + arena->last->size;
+    arena->last->size += size;
     return result;
+}
+
+void arena_clean(Arena *arena)
+{
+    for (Region *iter = arena->first;
+         iter != NULL;
+         iter = iter->next)
+    {
+        iter->size = 0;
+    }
+
+    arena->last = arena->first;
+}
+
+void arena_free(Arena *arena)
+{
+    Region *iter = arena->first;
+    while (iter != NULL) {
+        Region *next = iter->next;
+        free(iter);
+        iter = next;
+    }
+    arena->first = NULL;
+    arena->last = NULL;
+}
+
+void arena_summary(Arena *arena)
+{
+    if (arena->first == NULL) {
+        printf("[empty]");
+    }
+
+    for (Region *iter = arena->first;
+         iter != NULL;
+         iter = iter->next)
+    {
+        printf("[%zu/%zu] -> ", iter->size, iter->capacity);
+    }
+    printf("\n");
 }
 
 bool basm_resolve_binding(const Basm *basm, String_View name, Binding *binding)

--- a/src/bmr.c
+++ b/src/bmr.c
@@ -3,6 +3,46 @@
 
 #include <stdarg.h>
 
+typedef struct {
+    size_t size;
+    size_t capacity;
+    char *data;
+} Buffer;
+
+static void buffer_resize(Buffer *buffer, size_t new_capacity)
+{
+    buffer->capacity = new_capacity;
+    buffer->data = realloc(buffer->data, buffer->capacity);
+}
+
+static void buffer_write(Buffer *buffer, const char *data, size_t size)
+{
+    if (buffer->size + size > buffer->capacity) {
+        buffer_resize(buffer, buffer->capacity * 2 + size);
+    }
+
+    memcpy(buffer->data + buffer->size, data, size);
+    buffer->size += size;
+}
+
+static void buffer_free(Buffer *buffer)
+{
+    free(buffer->data);
+    buffer->data = NULL;
+    buffer->size = 0;
+    buffer->capacity = 0;
+}
+
+static String_View buffer_to_sv(Buffer *buffer)
+{
+    return (String_View) {
+        .count = buffer->size,
+        .data = buffer->data
+    };
+}
+
+static Buffer actual_buffer = {0};
+
 static void panic(const char *fmt, ...)
 {
     fprintf(stderr, "ERROR: ");
@@ -29,16 +69,6 @@ static void panic_errno(const char *fmt, ...)
     fprintf(stderr, ": %s\n", strerror(errno));
 
     exit(1);
-}
-
-static Arena actual_arena = {0};
-
-static String_View sv_from_arena(Arena *arena)
-{
-    return (String_View) {
-        .count = arena->size,
-        .data = arena->buffer,
-    };
 }
 
 static char *shift(int *argc, char ***argv)
@@ -75,8 +105,7 @@ static Err bmr_write(Bm *bm)
         return ERR_ILLEGAL_MEMORY_ACCESS;
     }
 
-    char *chunk = arena_alloc(&actual_arena, count);
-    memcpy(chunk, &bm->memory[addr], count);
+    buffer_write(&actual_buffer, (void*) &bm->memory[addr], count);
 
     bm->stack_size -= 2;
 
@@ -164,9 +193,8 @@ int main(int argc, char *argv[])
             panic_errno("could not save output to file `%s`", output_file);
         }
 
-        fwrite(actual_arena.buffer, sizeof(actual_arena.buffer[0]),
-               actual_arena.size,
-               output_file);
+        fwrite(actual_buffer.data, sizeof(actual_buffer.data[0]),
+               actual_buffer.size, output_file);
 
         if (ferror(output_file)) {
             panic_errno("could not save output to file `%s`", output_file);
@@ -183,11 +211,13 @@ int main(int argc, char *argv[])
                     expected_output_file_path, strerror(errno));
             exit(1);
         }
-        String_View actual_output = sv_from_arena(&actual_arena);
+        String_View actual_output = buffer_to_sv(&actual_buffer);
 
         compare_outputs(expected_output_file_path, expected_output, actual_output);
         printf("Expected output\n");
     }
+
+    buffer_free(&actual_buffer);
 
     return 0;
 }

--- a/src/bmr.c
+++ b/src/bmr.c
@@ -215,6 +215,8 @@ int main(int argc, char *argv[])
 
         compare_outputs(expected_output_file_path, expected_output, actual_output);
         printf("Expected output\n");
+
+        arena_free(&expected_arena);
     }
 
     buffer_free(&actual_buffer);

